shohrukh92/shop-be
# Task 9: Backend For Frontend

#### Task link: https://github.com/rolling-scopes-school/nodejs-aws-tasks/blob/main/task9-bff/task.md
#### Score: 7/7

#### Main tasks:
3 - A working and correct express application created in the bff-service folder.
5 - The bff-service was deployed with Elastic Beanstalk. The bff-service call is redirected to the appropriate service (_cart_ and _product_ services);

#### Products list API:
- BFF url: http://shohrukh92-bff-api-dev.eu-west-1.elasticbeanstalk.com/products
- BFF url (nestjs): http://shohrukh92-bff-nestjs-api-dev.eu-west-1.elasticbeanstalk.com/products
- API url: https://wdkwmyi86g.execute-api.eu-west-1.amazonaws.com/dev/products

#### Get Product by id:
- BFF url: http://shohrukh92-bff-api-dev.eu-west-1.elasticbeanstalk.com/products/7fc8513a-08d6-4960-95f0-23bcdec28ba7
- BFF url (nestjs): http://shohrukh92-bff-nestjs-api-dev.eu-west-1.elasticbeanstalk.com/products/7fc8513a-08d6-4960-95f0-23bcdec28ba7
- API url: https://wdkwmyi86g.execute-api.eu-west-1.amazonaws.com/dev/products/7fc8513a-08d6-4960-95f0-23bcdec28ba7

#### Get Product by id (error case):
- BFF url: http://shohrukh92-bff-api-dev.eu-west-1.elasticbeanstalk.com/products/123
- BFF url (nestjs): http://shohrukh92-bff-nestjs-api-dev.eu-west-1.elasticbeanstalk.com/products/123
- API url: https://wdkwmyi86g.execute-api.eu-west-1.amazonaws.com/dev/products/123

#### Get Profile cart:
- BFF url: http://shohrukh92-bff-api-dev.eu-west-1.elasticbeanstalk.com/cart
- BFF url (nestjs): http://shohrukh92-bff-nestjs-api-dev.eu-west-1.elasticbeanstalk.com/cart
- API url: https://tc4j0zofac.execute-api.eu-west-1.amazonaws.com/dev/api/profile/cart

#### 502 error case (when the service does not exist):
- BFF url: http://shohrukh92-bff-api-dev.eu-west-1.elasticbeanstalk.com/example
- BFF url (nestjs): http://shohrukh92-bff-nestjs-api-dev.eu-west-1.elasticbeanstalk.com/example
---

#### Additional tasks:
- +1 - Added a cache at the bff-service level for a request to the getProductsList function of the product-service. The cache expires in 2 minutes.
How to test:
&nbsp;&nbsp;&nbsp;&nbsp;Get products list: http://shohrukh92-bff-nestjs-api-dev.eu-west-1.elasticbeanstalk.com/products
&nbsp;&nbsp;&nbsp;&nbsp;Create a new product: https://d223ib3mk0bf19.cloudfront.net/admin/products
&nbsp;&nbsp;&nbsp;&nbsp;Get products list again - the result shouldn’t have a new product
&nbsp;&nbsp;&nbsp;&nbsp;Wait more than 2 minutes
&nbsp;&nbsp;&nbsp;&nbsp;Get products list again - the result should have a new product
- +1 - Created a NestJS version of bff-service based on express.js (bff-nestjs-service);
